File: lint-arwaky/archive/capabilities/config_rules_validator.py
"""config_rules_validator — Capability for validating and checking project configuration rules."""

from ..taxonomy import (
    AdapterName,
    AdapterStatus,
    BooleanVO,
    ProjectConfig,
    SuccessStatus,
)

from ..contract import IConfigRulesProtocol


class ConfigRulesValidator(IConfigRulesProtocol):
    """Business logic for interpreting and validating project configuration."""

    def __init__(self, project_config: ProjectConfig):
        self._config = project_config

    def is_adapter_enabled(self, adapter_name: AdapterName) -> SuccessStatus:
        """Determines if a specific adapter should run based on configuration rules."""
        for adapter in self._config.adapters:
            if adapter.name == adapter_name:
                return SuccessStatus(
                    value=BooleanVO(value=adapter.status == AdapterStatus.ENABLED)
                )

        # Default policy: enabled if not explicitly mentioned
        return SuccessStatus(value=BooleanVO(value=True))

    def validate_thresholds(self) -> SuccessStatus:
        """Validates that scoring thresholds are sane."""
        t = self._config.thresholds
        # Score must be 0-100
        if not (0 <= t.score.value <= 100):
            return SuccessStatus(value=BooleanVO(value=False))
        # Complexity and line limits must be positive
        if t.complexity.value <= 0 or t.max_file_lines.value <= 0:
            return SuccessStatus(value=BooleanVO(value=False))
        return SuccessStatus(value=BooleanVO(value=True))

File: lint-arwaky/src-rust/capabilities/config_rules_validator.rs
// config_rules_validator — Capability for validating project configuration rules.
// Implements IConfigRulesProtocol: is_adapter_enabled, validate_thresholds.

use crate::taxonomy::{AdapterName, ProjectConfig, AdapterStatus};

/// Result of a validation operation.
pub struct ValidationResult {
    pub is_valid: bool,
    pub reason: Option<String>,
}

impl ValidationResult {
    pub fn ok() -> Self {
        Self { is_valid: true, reason: None }
    }
    pub fn fail(reason: &str) -> Self {
        Self { is_valid: false, reason: Some(reason.to_string()) }
    }
}

/// Business logic for interpreting and validating project configuration.
pub struct ConfigRulesValidator {
    config: ProjectConfig,
}

impl ConfigRulesValidator {
    pub fn new(config: ProjectConfig) -> Self {
        Self { config }
    }

    /// Determines if a specific adapter should run based on configuration rules.
    pub fn is_adapter_enabled(&self, adapter_name: &AdapterName) -> bool {
        for adapter in &self.config.adapters {
            if adapter.name.value == adapter_name.value {
                return adapter.status == AdapterStatus::ENABLED;
            }
        }
        // Default policy: enabled if not explicitly mentioned
        true
    }

    /// Validates that scoring thresholds are sane.
    pub fn validate_thresholds(&self) -> ValidationResult {
        let t = &self.config.thresholds;

        // Score must be 0-100
        if !(0.0..=100.0).contains(&t.score.value) {
            return ValidationResult::fail("Score threshold must be between 0 and 100.");
        }

        // Complexity and line limits must be positive
        if t.complexity.value <= 0 {
            return ValidationResult::fail("Complexity threshold must be positive.");
        }

        if t.max_file_lines.value <= 0 {
            return ValidationResult::fail("max_file_lines threshold must be positive.");
        }

        ValidationResult::ok()
    }
}
